Add QgsProcessingFeatureBasedAlgorithm subclass

nyalldawson opened the pull request:
Now that we've got a number of algorithms ported successfully to the new API (and things are looking stable and ready to be locked in there), it's time for (hopefully) the final substantial addition to the new processing backend.

This PR introduces "QgsProcessingFeatureBasedAlgorithm". This is an abstract QgsProcessingAlgorithm base class for processing algorithms which operate "feature-by-feature".

Feature based algorithms are algorithms which operate on individual features in isolation. These are algorithms where one feature is output for each input feature, and the output feature result for each input feature is not dependent on any other features present in the source.

For instance, algorithms like "centroids" and "buffers" are feature based algorithms since the centroid or buffer of a feature is calculated for each feature in isolation. An algorithm like "dissolve" is NOT suitable for a feature based algorithm as the dissolved output depends on multiple input features and these features cannot be processed in isolation.

Using QgsProcessingFeatureBasedAlgorithm as the base class for feature based algorithms allows shortcutting much of the common algorithm code for handling iterating over sources and pushing features to output sinks.

It also allows the algorithm execution to be optimised in future. Potential future benefits for QgsProcessingFeatureBasedAlgorithm are:
- allowing automatic multi-thread processing of the algorithm
- allowing use of the algorithm in "pipes", avoiding the need for temporary outputs in multi-step models and potentially giving a huge speed boost to models (e.g. instead of a model reprojecting a layer, and then buffering the resulting layer, features could instead be piped from the original source, reprojected, buffered, and ONLY then saved to an output layer. End result: LOTS less memory overhead, CPU cooking.). 
- allowing some feature based algorithms to operate "in place" (make a selection in an editable layer, run the algorithm over just those selected features and it modifies them in place, without creating a new layer. At the moment these kinds of editing operations are a PITA - you have to create a new temporary layer with the result, then delete features from the original layer and copy and paste the features from the temporary layer back into the main layer. No thanks!) Only algorithms which keep the same WKB type, CRS, and fields would be suitable for this.

nyalldawson committed:
Add QgsProcessingFeatureBasedAlgorithm subclass

An abstract QgsProcessingAlgorithm base class for processing algorithms
which operate "feature-by-feature".

Feature based algorithms are algorithms which operate on individual
features in isolation. These are algorithms where one feature is
output for each input feature, and the output feature result
for each input feature is not dependent on any other features
present in the source.

For instance, algorithms like "centroids" and "buffers" are feature
based algorithms since the centroid or buffer of a feature is
calculated for each feature in isolation. An algorithm like "dissolve"
is NOT suitable for a feature based algorithm as the dissolved output
depends on multiple input features and these features cannot be
processed in isolation.

Using QgsProcessingFeatureBasedAlgorithm as the base class for feature
based algorithms allows shortcutting much of the common algorithm code
for handling iterating over sources and pushing features to output sinks.
It also allows the algorithm execution to be optimised in future
(for instance allowing automatic multi-thread processing of the
algorithm, or use of the algorithm in "chains", avoiding the need
for temporary outputs in multi-step models).

nyalldawson committed:
Port some existing algorithms to QgsProcessingFeatureBasedAlgorithm

nyalldawson committed:
Port a single python algorithm to QgsProcessingFeatureBasedAlgorithm

nyalldawson committed:
[FEATURE] New algorithms to add Z/M values to existing geometries

Allows upgrading geometries to include these dimensions, or
overwriting any existing Z/M values with a new value.

Intended mostly as a test run for QgsProcessingFeatureBasedAlgorithm

nirvn commented:
> allowing some feature based algorithms to operate "in place"

That's really cool.

nyalldawson committed:
[FEATURE] Followup addition of set Z/M values algs with Drop Z/M Values algorithm

Allows easy access to drop any z or m values present in a layer (e.g. if
required for compatibility with a database destination, etc)

alexbruy commented:
Cool!

m-kuhn commented on src/core/processing/qgsprocessingalgorithm.h:
```diff
@@ -697,6 +699,115 @@ Q_DECLARE_OPERATORS_FOR_FLAGS( QgsProcessingAlgorithm::Flags )
+     * output WKB type generated by the algorithm. The default behavior is that the algorithm maintains
+     * the same WKB type.
+     */
+    virtual QgsWkbTypes::Type outputWkbType( QgsWkbTypes::Type inputWkbType ) const { return inputWkbType; }
```
What's that used for? Just wondering how that integrates with, say `createFromAttributeWkt` which also operates feature-by-feature.

nyalldawson replied:
What's `createFromAttributeWkt`? It's used by the base class to determine the output wkb type during creation of the sink. It requires the input wkb type for algorithms like Add M Value or Boundary which need to transform certain input WKB types to other WKB output types.

m-kuhn replied:
Ah, that's an initial step, not a per-feature one. In this case one could override it in a subclass and return `GeometryCollection` or allow configuration for such an algorithm as `createFromWkt`, all good.

nyalldawson replied:
Ok - looks like I need to improve the dox then!

nyalldawson replied:
Fixed

m-kuhn commented on src/core/processing/qgsprocessingalgorithm.cpp:
```diff
@@ -608,3 +608,67 @@ bool QgsProcessingAlgorithm::createAutoOutputForParameter( QgsProcessingParamete
+
+  long count = mSource->featureCount();
+  if ( count <= 0 )
+    return QVariantMap();
```
How does that behave in a model with an empty layer? Will the next algorithm not complain that it wasn't provided with a layer (while IMO the expected behavior is that it gets provided with an empty layer)?

nyalldawson replied:
Good point! I'll fix.

nyalldawson replied:
Fixed

m-kuhn commented:
Great work. Nice idea :)

> These are algorithms where one feature is output for each input feature

Why this requirement? It sounds to me like an artificial restriction.

What do you think of talking about a "feature stream"? A QgsFeatureSource provides a feature stream, a QgsFeatureThink consumes a feature stream.

Then, a feature filter algorithm could take a feature stream as input and produce a feature stream as output and still use the infrastructure here.

nyalldawson commented:
> What do you think of talking about a "feature stream"? A QgsFeatureSource provides a feature stream, a QgsFeatureThink consumes a feature stream.

Isn't that just a feature iterator by a different name?

> > These are algorithms where one feature is output for each input feature
> Why this requirement? It sounds to me like an artificial restriction.

Are you thinking returning a list of features here? That could be useful for some algs I guess.

m-kuhn commented:
> Isn't that just a feature iterator by a different name?

Totally

>  Are you thinking returning a list of features here? That could be useful for some algs I guess.

In python terms I was thinking about a *generator* rather than an *iterator* (with the obvious advantage memory wise). I was just wondering if you had a good reason for the *one feature in, one feature out* restriction (like piping a feature stream through two algorithms in parallel and joining them later or so).

nyalldawson commented:
@m-kuhn gitterize!

nyalldawson committed:
Always output algorithm results, even when input source has no
features

Allows more versatile models which can handle empty layers

nyalldawson committed:
Improve dox

nyalldawson commented:
I've chatted with @m-kuhn about this offline - the outcome was that we could later extend this to allow outputting multiple features per input feature via something like

    virtual QgsFeatureList processFeatures( QgsFeatureList features )
    {
        result = QgsFeatureList()
        for feature in features:
            if ( self.processFeature( feature ) )
                result.append( self.processFeature( feature ) )

        return result
    }

in the QgsProcessingFeatureBasedAlgorithm class. But that's not an API break so we can address that later if needed.

m-kuhn commented:
> In the QgsProcessingFeatureBasedAlgorithm class. But that's not an API break so we can address that later if needed.

No? I would have imagined it like

```
class QgsProcessingFeatureBasedAlgorithm
{
    public:
      virtual QgsFeatureList processFeatures( const QgsFeatureList& features );

    private:
      virtual QgsFeature processFeature( const QgsFeature& feature );
}
```

vs the current approach

```
class QgsProcessingFeatureBasedAlgorithm
{
    public:
      virtual QgsFeature processFeature( const QgsFeature& feature );
}
```

nyalldawson commented:
Unless I'm missing something, that's not an API break...?

nyalldawson commented:
Oh wait - you're meaning if someone starts calling processFeature(...) on the alg. I was thinking in terms of subclassing.

Do private virtual's even work with SIP? I had no success last time I tried these....

m-kuhn commented:
> Do private virtual's even work with SIP? I had no success last time I tried these....

I think that has been addressed meanwhile but don't ask me what makes me think that :D 

In any case, being transparent about the fact that this method should only be reimplemented and that one should not rely upon direct calls should be fine.

nyalldawson commented:
@m-kuhn 

Here's what I'm thinking now:

- I'm not a fan of *forcing* QgsProcessingFeatureBasedAlgorithm to use a list for features. I'm concerned about:
    - the extra overhead here - instead of just altering features in place, we're creating a bunch of lists and passing features to and from using them. That's a lot of extra allocations that we can hopefully avoid. 
    - the extra complexity  - we'd need to work out a good batch size to send features to be processed by the algorithm. This could get tricky - some algs will perform their calculations almost instantly, others may take seconds/minutes per feature.
    - responsiveness - we'd need to pass the QgsFeedback object to processFeatures so that the operation can be canceled mid way through processing the list (otherwise we'd lose responsive cancelation/progress reports for algorithms which take a long time to process each feature).

- I'd rather have a second QgsProcessingAlgorithm subclass (QgsProcessingMultiFeatureBasedAlgorithm?) which behaves in the way you've proposed - via a `virtual QgsFeatureList processFeatures( const QgsFeatureList& features );` method. Algs which return multiple features for a single input feature would derive from this class instead.
- We could add a base (non virtual, not called by processAlgorithm) processFeatures implementation to QgsProcessingFeatureBasedAlgorithm which just calls processFeature on each feature in the list in turn. Then if needed we could have an interface so that code which uses processFeatures doesn't need to care whether or not the alg is a QgsProcessingFeatureBasedAlgorithm or a QgsProcessingMultiFeatureBasedAlgorithm.

Does that sound reasonable?

m-kuhn commented:
Yes, sounds reasonable

> the extra overhead here - instead of just altering features in place, we're creating a bunch of lists and passing features to and from using them. That's a lot of extra allocations that we can hopefully avoid.

then why not:

    void processFeature( QgsFeature &feature );

though it will be more complicated when the same output feature is routed to multiple inputs of other algs in parallel

nyalldawson commented:
virtual bool processFeature( QgsFeature &feature, QgsProcessingFeedback *feedback ) = 0;

m-kuhn commented:
What happens if false is returned?

Just wondered if ref-to-feature instead of const-ref even helps a lot when we still will have plenty of QgsGeometry detaches with every alg.

nyalldawson commented:
> What happens if false is returned?

The feature is discarded and not pushed to the sink. 

> Just wondered if ref-to-feature instead of const-ref even helps a lot when we still will have plenty of QgsGeometry detaches with every alg.

Probably not - but does it hurt?

m-kuhn commented:
> The feature is discarded and not pushed to the sink.

Ok, I was wondering about the other possibility of completely failing the job, but that should be sent through feedback in this case?

> Probably not - but does it hurt?

Not really. Would it help to do this though:

```
    geometry = feature.takeGeometry()
    geometry.modify()
    feature.setGeometry( geometry )
```

nyalldawson commented:
> Ok, I was wondering about the other possibility of completely failing the job, but that should be sent through feedback in this case?

You can give individual logging events through feedback (like GEOS failures), but to completely halt the algorithm (and model) you'd throw a QgsProcessingException. (I'll add that to the docs)

I'll update the signature to use

    virtual QgsFeature processFeature( const QgsFeature &feature, QgsProcessingFeedback *feedback ) = 0;

ultimately I don't think either way would have significant performance implications. Do you think returning an invalid feature should be treated as "skip" this feature?

nyalldawson committed:
Add note about throwing QgsProcessingException to cancel algorithm execution

nyalldawson committed:
Change signature of processFeature so that features are no longer modified in place

m-kuhn commented:
> Do you think returning an invalid feature should be treated as "skip" this feature?

Probably, I don't see anything else that makes sense (after your last comment about failing the job is done with an exception).

nyalldawson committed:
Skip invalid returned features

nyalldawson commented:
@m-kuhn all good to go for you now?

m-kuhn reviewed (approved):
Nice, good work!
